tools/dotnet-linker/Steps/ManagedRegistrarLookupTablesStep.cs

```diff
@@ -0,0 +1,414 @@
+			il.Emit (OpCodes.Ldloca, temporary);
+			il.Emit (OpCodes.Initobj, abr.System_RuntimeTypeHandle);
+			il.Emit (OpCodes.Ldloc, temporary);
+			il.Emit (OpCodes.Ret);
+		}
+
+		void GenerateRegisterWrapperTypes (TypeDefinition type)
+		{
+			var method = type.AddMethod ("RegisterWrapperTypes", MethodAttributes.Private | MethodAttributes.Final | MethodAttributes.Virtual | MethodAttributes.NewSlot | MethodAttributes.HideBySig, abr.System_Void);
+			var git = new GenericInstanceType (abr.System_Collections_Generic_Dictionary2);
+			git.GenericArguments.Add (abr.System_RuntimeTypeHandle);
+			git.GenericArguments.Add (abr.System_RuntimeTypeHandle);
+			method.AddParameter ("type", git);
+			method.Overrides.Add (abr.IManagedRegistrar_RegisterWrapperTypes);
+			method.CreateBody (out var il);
+
+			var addMethodReference = abr.System_Collections_Generic_Dictionary2.CreateMethodReferenceOnGenericType (abr.Dictionary2_Add, abr.System_RuntimeTypeHandle, abr.System_RuntimeTypeHandle);
+			var currentTypes = StaticRegistrar.Types.Where (v => v.Value.Type.Resolve ().Module.Assembly == abr.CurrentAssembly);
+			foreach (var ct in currentTypes) {
+				if (!ct.Value.IsProtocol)
+					continue;
+				if (ct.Value.ProtocolWrapperType is null)
+					continue;
+
+				var keyMarked = !IsLinkedAway (ct.Key.Resolve ());
+				var wrapperTypeMarked = !IsLinkedAway (ct.Value.ProtocolWrapperType.Resolve ());
+				if (!keyMarked && !wrapperTypeMarked)
+					continue;
+				if (keyMarked ^ wrapperTypeMarked)
+					throw new InvalidOperationException ($"Huh?");
+
+				il.Emit (OpCodes.Ldarg_1);
+				il.Emit (OpCodes.Ldtoken, type.Module.ImportReference (ct.Key));
+				il.Emit (OpCodes.Ldtoken, type.Module.ImportReference (ct.Value.ProtocolWrapperType));
+				il.Emit (OpCodes.Call, addMethodReference);
+			}
+
+			il.Emit (OpCodes.Ret);
+		}
+
+		void GenerateLookupUnmanagedFunction (TypeDefinition registrar_type, IList<TrampolineInfo> trampolineInfos)
+		{
+			Console.WriteLine ($"GenerateLookupMethods ({registrar_type.FullName}, {trampolineInfos.Count} items");
+
+			MethodDefinition? lookupMethods = null;
+			if (App.IsAOTCompiled (abr.CurrentAssembly.Name.Name)) {
+				// Don't generate lookup code, because native code will call the EntryPoint for the UnmanagedCallerOnly methods directly.
+				Console.WriteLine ($"Not generating method lookup code for {abr.CurrentAssembly.Name.Name}, because it's AOT compiled");
+			} else if (trampolineInfos.Count > 0) {
+				// All the methods in a given assembly will have consecutive IDs (but might not start at 0).
+				if (trampolineInfos.First ().Id + trampolineInfos.Count - 1 != trampolineInfos.Last ().Id)
+					throw ErrorHelper.CreateError (99, $"Invalid ID range: {trampolineInfos.First ().Id} + {trampolineInfos.Count - 1} != {trampolineInfos.Last ().Id}");
+
+				const int methodsPerLevel = 10;
+				var levels = (int) Math.Ceiling (Math.Log (trampolineInfos.Count, methodsPerLevel));
```

Comment:
This will likely produce incorrect result if `trampolineInfos.Count == 1`. The logarithm ends up being 0, and so is `levels`.

Reply:
```suggestion
				var levels = (int) Math.Ceiling (Math.Log (trampolineInfos.Count, methodsPerLevel));
				levels = levels == 0 ? 1 : levels;
```